wpimath/src/main/java/edu/wpi/first/math/estimator/PoseEstimator.java

```diff
@@ -120,7 +118,13 @@ public void resetPosition(Rotation2d gyroAngle, T wheelPositions, Pose2d poseMet
    * @return The estimated robot pose in meters.
    */
   public Pose2d getEstimatedPosition() {
-    return m_odometry.getPoseMeters();
+    return m_poseEstimate;
+  }
+
+  public Optional<Pose2d> sampleAt(double timestamp) {
+    return m_poseBuffer
+        .getSample(timestamp)
+        .map(sample -> sample.exp(m_odometry.getPoseMeters().log(m_poseEstimate)));
```

Comment:
Is this supposed to be different from the logic for `old_estimate` in `addVisionMeasurement`? `a.exp(b.loc(c)) = a + (c - b)`, which does not necessarily equal `c + (a - b)`, so the logic here doesn't necessarily produce the same result as the calculation in `addVisionMeasurement`. (This method also does not seem to be used at all, in library code or test code)

Quick sketch of why `a.exp(b.log(c)) = a + (c - b)`: From looking at the code, `pose.exp(twist) = pose + f(twist)` and `pose₁.log(pose₂) = g(pose₂ - pose₁)`. Because `c.exp(c.log(b)) = b` for all `c` and `b`, then `c + f(g(c - a)) = b` for all `c` and `b`, and since the only value `x` satisfying `c + x = b` is `x = b - c`, then `f(g(c - a)) = c - a` for all `c` and `b`. However, if we expand `a.exp(b.log(c))`, we get `a + f(g(c - b))`, which we know is `a + (c - b)` from what we just proved.

However, `a + (c - b)` is not necessary equal to `c + (a - b)`. Letting `T(pose)` and `T(transform)` be the translation component of the pose or transform, respectively, letting `R(pose)` be the rotation component of the pose, and letting `rot(t, r)` be the result of rotating translation `t` by rotation `r`, then:
```
T(a + (c - b)) = T(a) + rot(T(c - b), R(a))
T(a + (c - b)) = T(a) + rot(rot(T(c) - T(b), -R(b), R(a))
T(a + (c - b)) = T(a) + rot(T(c) - T(b), R(a) - R(b))
T(a + (c - b)) = T(a) + rot(T(c), R(a) - R(b)) - rot(T(b), R(a) - R(b))
```
From this, we can see that `T(a + (c - b)) ≠ T(c + (a - b)) iff R(a) ≠ R(b)`.

Reply:
I couldn't understand your proof of `a.exp(b.log(c)) = a + (c - b)`, so I wrote a different one. Let a, b, and c be elements of SE(3).
```
a.exp(b.log(c))
```
This code is equivalent to the following operators:
```
a + exp(log(c - b))
```
where `b.log(c)` corresponds to `log(c - b)` and `a.exp(d)` corresponds to `a + exp(d)`. We've defined + and - to use transformations instead of element-wise addition and subtraction in this context.

Since `exp()` and `log()` are inverse operators, we get
```
a + (c - b)
```

Note that the same does not hold for interpolation. That is, `a + exp(t⋅log(c - b)) ≠ a + t⋅(c - b)` where t∈(0, 1). Since exp() is not a linear operator, t can't be factored out to make exp() and log() cancel.

If we sweep t, the left-hand side looks like tracing an arc while the right-hand side is linear interpolation.

Reply:
Based on the above comments I've updated `sampleAt` to use transforms instead of twists, since the functionality is equivalent but the former is less computationally expensive.